### PR TITLE
carbonzipper: introduce deterministicBackend option

### PR DESCRIPTION
**Pull request by bom-d-van**
## What issue is this change attempting to solve?

In graphite storage (go-carbon) installations where consistency can not be guaranteed,
by always prefer to return metrics from the same set of backends would make the query
result less confusing, no more flapping results.

This commit essentially make graphite render result more consistent, both and right
and wrong.

## How does this change solve the problem? Why is this the best approach?

DeterministicBackend sets a backend weight for each metrics in render
request and carbonzipper would favor metrics from a specific node.
As storage backends like go-carbon does not have strong consistency
guarantee, this means that the same metric persisted in different
replicas might have different values. Without DeterministicBackend
option, carbonzipper would just favor metrics returned first from
any backend. This might result in confusing query result in
carbonapi that are reading multiple metrics in one query.

An example like bellow:

```
asPercent(
    integral(sumSeries(http.app.latencies.{0_5_1,1_5,5_10,10_25,25_50,50_100,100_250}.count)),
    integral(sumSeries(http.app.latencies.*.count))
)
```

Without DeterministicBackend option, the example query above might
return result 100.1%. With DeterministicBackend option, it is
guaranteed that the query result would always "seems" meaningful,
but not necessarily correct.

## How can we be sure this works as expected?

Check TestMergeMetricsDifferingStepTimes7. Also confirm with production data that returning results are always consistent now.

### CODE REVIEW

File: cfg/common.go [hunk 1]
```diff
@@ -167,6 +167,28 @@ type Common struct {
 
 	Traces               Traces `yaml:"traces"`
 	PrintErrorStackTrace bool   `yaml:"printErrorStackTrace"`
+
+	// DeterministicBackend sets a backend weight for each metrics in render
```
Comment (timtofan) [inline cfg/common.go:171]: I think it should be either `each metric` or `all metrics`.

File: cfg/common.go [hunk 2]
```diff
@@ -167,6 +167,28 @@ type Common struct {
...
+	// An example like bellow:
+	//
+	//     asPercent(
+	//         integral(sumSeries(http.app.latencies.{0_5_1,1_5,5_10,10_25,25_50,50_100,100_250}.count)),
+	//         integral(sumSeries(http.app.latencies.*.count))
+	//     )
+	//
+	// Without DeterministicBackend option, the example query above might
+	// return result 100.1%. With DeterministicBackend option, it's
+	// guaranteed that the query result would always "seems" meaningful,
```
Comment (timtofan) [inline cfg/common.go:189]: result would `seems` vs result would `seem`

File: pkg/types/types.go [hunk 3]
```diff
@@ -231,7 +239,21 @@ func (s byStepTime) Swap(i, j int) {
 }
 
 func (s byStepTime) Less(i, j int) bool {
-	return s[i].StepTime < s[j].StepTime
+	// no deterministic backend sorting
+	if s[i].StepTime != s[j].StepTime || s[i].backendAddr == "" {
+		return s[i].StepTime < s[j].StepTime
+	}
+
+	if s[i].backendWeight < s[j].backendWeight {
```
Comment (timtofan) [inline pkg/types/types.go:247]: What's the advantage of comparing metricName+address hashes as opposed to just checking addresses as happens in `// dealing with hash collision`?

File: pkg/types/types.go [hunk 4]
```diff
@@ -231,7 +239,21 @@ func (s byStepTime) Swap(i, j int) {
 }
 
 func (s byStepTime) Less(i, j int) bool {
```
Comment (timtofan) [inline pkg/types/types.go:241]: I think we can use sort.Slice instead of special type for sorting. It [was introduced in go 1.8](https://go.dev/doc/go1.8#sort_slice). OTOH maybe it should be a separate patch 🤷

File: pkg/types/types.go (hunk 3 above)
Comment (bom-d-van) [reply pkg/types/types.go:247]: the downside of just sorting by server address is that we might potentially always return metrics in a dc/az due to it's hostnames or ip addresses.
I think it's better to uniformly retrieve metrics from different replicas but still stick to the same server in every api calls.

File: pkg/types/types.go (hunk 4 above)
Comment (bom-d-van) [reply pkg/types/types.go:241]: yep. could be a separate change. and as a separate function also isn't bad, we can reuse it somewhere else if needed.

File: cfg/common.go [hunk 5]
```diff
@@ -167,6 +167,28 @@ type Common struct {
...
+	//
+	// An example like bellow:
+	//
+	//     asPercent(
+	//         integral(sumSeries(http.app.latencies.{0_5_1,1_5,5_10,10_25,25_50,50_100,100_250}.count)),
+	//         integral(sumSeries(http.app.latencies.*.count))
+	//     )
+	//
+	// Without DeterministicBackend option, the example query above might
+	// return result 100.1%. With DeterministicBackend option, it's
```
[the remark this answers is not in the document]
Comment (bom-d-van) [reply cfg/common.go:188]: Not for this example. Reword the comments to make it clear:

> Without DeterministicBackend option, if there are inconsistent values
> between replicas in the returning metrics (for example, in one
> replica 0_5_1.count has value 10, and in another 0_5_1.count has 5),
> the example query above might return percentage values larger than
> 100%. With DeterministicBackend option, it's guaranteed that the
> query result would always "seem" meaningful (in the example, it
> always return value <= 100%), but not necessarily correct.

File: cfg/common.go (hunk 1 above)
Comment (bom-d-van) [reply cfg/common.go:171]: thanks, fixed now.

File: cfg/common.go (hunk 2 above)
Comment (bom-d-van) [reply cfg/common.go:189]: thanks, fixed now.

File: cfg/common.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (timtofan) [reply cfg/common.go]: Hmm, sorry, after rephrase it didn't get any clearer for me 😬

File: pkg/types/types.go (hunk 3 above)
Comment (timtofan) [reply pkg/types/types.go:247]: I see. I think given it's not obvious, maybe worth adding a comment.

Would it be sufficient to hash address w/ metric name to achieve the same result?

Comment (bom-d-van) [reply pkg/types/types.go:247]: > I see. I think given it's not obvious, maybe worth adding a comment.

yep, sounds good.

> Would it be sufficient to hash address w/o the metric name to achieve the same result?

then it's just the same as sorting backend address?

Comment (bom-d-van) [reply pkg/types/types.go:248]: Adds the comment here: https://github.com/bookingcom/carbonapi/pull/340/files#diff-9b4292c30cc53778b4e267af519527527024f6c158979cb0a0a55ca0192651a0R162-R165

File: cfg/common.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (bom-d-van) [reply cfg/common.go]:
```
asPercent(
    integral(sumSeries(http.app.latencies.{0_5_1,1_5,5_10,10_25,25_50,50_100,100_250}.count)),
    integral(sumSeries(http.app.latencies.*.count))
)
```

Let's suppose there are two replicas and only metric `0_5_1.count` has inconsistent values at `timestamp T` (`10` in replica a, `5` in replica b).

Without `DeterministicBackend`, carbonzipper might fetches `0_5_1.count` for `integral(sumSeries(http.app.latencies.{0_5_1,1_5,5_10,10_25,25_50,50_100,100_250}.count)),` from replica a, and fetches `0_5_1.count` for `integral(sumSeries(http.app.latencies.*.count))` from replica b.

These means that at `timestamp T`, `integral(sumSeries(http.app.latencies.{0_5_1,1_5,5_10,10_25,25_50,50_100,100_250}.count))` is larger than `integral(sumSeries(http.app.latencies.*.count))`, and it would produces a percentage value >= 100%.

With `DeterministicBackend`, carbonzipper would either use replica a or replica b for both subqueries thus its guarantee to return a percentage value <= 100%, which is arguably more meaningful.

**Review by grzkv — DISMISSED** [review]
@bom-d-van If I understand everything correctly, I am not sure this adds value to the system. Consistently returning a wrong result could be considered worse than inconsistently returning results. Inconsistency still is an indicator to the user - even if awkward - that something is not right.

Surely, it is better to let the user know that the data is wrong. This change hides the errors the system makes. It makes bad data look good.

File: cfg/common.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (timtofan) [reply cfg/common.go]: ohh, yep got it now, thx

File: pkg/types/types.go (hunk 3 above)
Comment (timtofan) [reply pkg/types/types.go:248]: > then it's just the same as sorting backend address?
true, nvm

File: pkg/backend/net/net.go [hunk 6]
```diff
@@ -154,6 +159,16 @@ func (b Backend) GetServerAddress() string {
 	return b.address
 }
 
+// Motivations of producing a weight by hashing metric_name + server_addr:
+//
+// 	Make carbonzipper uniformly retrieving metrics from different
```
Comment (timtofan) [inline pkg/backend/net/net.go:164]: `retrieving` vs `retrieve`

File: pkg/backend/net/net.go [hunk 7]
```diff
@@ -154,6 +159,16 @@ func (b Backend) GetServerAddress() string {
...
+// 	replicas but still stick to the same backend server in every api calls.
```
Comment (timtofan) [inline pkg/backend/net/net.go:165]: `every api calls` vs `every api call`

File: pkg/backend/net/net.go (hunk 6 above)
Comment (bom-d-van) [reply pkg/backend/net/net.go]: 🤦 thanks.
